### ISSUE TITLE
Backface-cull some water source tiles to reduce z-fighting

### ISSUE THREAD

**Issue by paramat** [minetest/minetest_game#2238]
As a solution for water z-fighting with nodeboxes, previously engine issue https://github.com/minetest/minetest/issues/3704
@Fixer-007 

With all tiles backface-culled:

![screenshot_20181012_013311](https://user-images.githubusercontent.com/3686677/46841477-17b5b780-cdc0-11e8-9082-7e8d6d0a0260.png)

Works but now the water surface is not visible from underneath.
Here's unchanged code:

![screenshot_20181012_013418](https://user-images.githubusercontent.com/3686677/46841497-2f8d3b80-cdc0-11e8-8e58-a5613dda04b8.png)

Now with all water source tiles except the top one backface culled:

![screenshot_20181012_013609](https://user-images.githubusercontent.com/3686677/46841509-4469cf00-cdc0-11e8-8664-257df4860930.png)

Removes most visible z-fighting (the undersides of nodeboxes are less commonly seen and the darkness somewhat hides the glitch) and preserves water surface being visible.
Remaining glitch:

![screenshot_20181012_014114](https://user-images.githubusercontent.com/3686677/46841562-b2ae9180-cdc0-11e8-8deb-3bedf741a3b5.png)

Suggestion for water source:
```
	tiles = {
		{
			name = "default_water_source_animated.png",
			backface_culling = false,
			animation = {
				type = "vertical_frames",
				aspect_w = 16,
				aspect_h = 16,
				length = 2.0,
			},
		},
		{
			name = "default_water_source_animated.png",
			backface_culling = true,
			animation = {
				type = "vertical_frames",
				aspect_w = 16,
				aspect_h = 16,
				length = 2.0,
			},
		},
	},
```
Note that flowing water already does this:
```
	special_tiles = {
		{
			name = "default_water_flowing_animated.png",
			backface_culling = false,
			animation = {
				type = "vertical_frames",
				aspect_w = 16,
				aspect_h = 16,
				length = 0.8,
			},
		},
		{
			name = "default_water_flowing_animated.png",
			backface_culling = true,
			animation = {
				type = "vertical_frames",
				aspect_w = 16,
				aspect_h = 16,
				length = 0.8,
			},
		},
	},
```
Just needs some testing to check for any unforseen problems.

**Comment by paramat** [comment]
Looked through MTG history, these backface culling parameters have been this way for source and flowing since 2012 https://github.com/minetest/minetest_game/commit/d817c9babeedb335735f3a9c7efb4d325da27eb9 . No PR thread, it was directly commited by c55.
So far i can't see a reason for water source to be this way. This was probably before nodeboxes too.

**Comment by paramat** [comment]
> So for water source it will be.

No. You no longer see the water surface at y = 1 with all tiles backface culled.
'special tiles' for water source is possibly no longer used as 'new style water' was (i think) removed from the engine, i'll check that.

**Comment by TumeniNodes** [comment]
definitely best to check, as I know when I fiddled with everything, changes to it caused a blank face (meaning coloured) to indicate no texture found.
I can try to explain better if needed

**Comment by paramat** [comment]
'special tiles' was added to liquid sources in fd1135c7af46eb2f5b99a11f48bf9f9ae335ea9c and b3786d84c5a7dc709c7b2a87fcd472013484b9e9 still not sure where and when they are/were used, @numberZero any idea?

**Comment by paramat** [comment]
Removing 'special_tiles' from water source seems to have no effect.

**Comment by TumeniNodes** [comment]
I tried that as well.
If you then dig a small ditch, and it sets water flowing, and then block it with a node, then dig the node... you will see the issue.

I wondered why both, identical defs are used, and found out why but, I still do not understand it.

**Comment by paramat** [comment]
Ok, we can leave 'special_tiles' unchanged for now then.

**Comment by Fixer-007** [comment]
I'm not sure about this, maybe someone who can into graphics can help?

**Comment by paramat** [comment]
That's why i tagged numberZero. The relevant commits are linked above, June 2012, possibly pre-RBA.

**Comment by paramat** [comment]
However, the need for 'special tiles' is not a problem, we can leave that in, the suggested changes work and are straightforward backface culling stuff.

**Comment by numberZero** [comment]
@paramat This way, being inside a waterfall you won’t see its surface.

> 'special tiles' was added to liquid sources in fd1135c and b3786d8 when liquid animations were added. Still not sure where they are or were used, @numberZero any idea?

Currently liquid sources use normal tiles only (as they are drawn just like normal nodes), and flowing liquid uses special tiles only. The only reason I may see here as flowing liquid’s tiles are not aligned in any way and are oriented as the engine likes.

BTW, `special_tiles` is what was `special_materials` before fd1135c. It only introduced animation.

**Comment by paramat** [comment]
> being inside a waterfall you won’t see its surface.

Good point, i'm uncertain then.

**Comment by Ezhh** [comment]
This seems to be another "make one thing better and one thing worse" type of trade off. Losing the ability to see the surface while in flowing liquid sounds like a real issue, and this also isn't a full fix, so I don't think we should do it.

**Comment by paramat** [comment]
Agreed.

**Comment by paramat** [comment]
Actually, a waterfall or water elevator consists of flowing nodes, so this change wouldn't make any difference. Will test anyway.

**Comment by HybridDog** [comment]
Is it possible to add a 0.0001f offset to the faces in minetest core

**Comment by paramat** [comment]
Yes, but not a good idea, and not necessary.

**Comment by numberZero** [comment]
> Yes, but not a good idea, and not necessary.

Not that easy, actually. Water is drawn using fast-faces, these aren’t easily tweak-able. Also that would cause the same broken edges as in https://github.com/minetest/minetest/pull/7514#issuecomment-422945774 (caused by shader misbehavior there). Nevertheless, it’s possible to use shaders to alter depth slightly, thus giving either translucent or opaque faces the priority.

**Comment by paramat** [comment]
Oddly enough, when inside a water elevator, you already can't see the edges from inside, the suggested change makes no difference (as expected, as a water elevator or waterfall is all flowing nodes):

![screenshot_20181019_101027](https://user-images.githubusercontent.com/3686677/47209235-be352600-d387-11e8-8936-15539730cd3d.png)

**Comment by paramat** [comment]
Now if i set all flowing tiles to have no backface culling, you can see the sides of a waterfall from inside:

![screenshot_20181019_103346](https://user-images.githubusercontent.com/3686677/47210403-c04cb400-d38a-11e8-9f83-f2f12debdcdb.png)

However from outside it now looks like this:

![screenshot_20181019_103455](https://user-images.githubusercontent.com/3686677/47210422-ccd10c80-d38a-11e8-89e6-822e15b27e54.png)

**Comment by paramat** [comment]
> BTW, special_tiles is what was special_materials before fd1135c

Thanks just seen what imissed.
I traced water source special_materials and 'new style water' back to 24 May 2011:
https://github.com/minetest/minetest/blob/99c2ac0125f9bbae0a4b06e6e7145f76f4dfbbf4/src/mapnode.cpp#L281
```
	if(new_style_water)
	{
		f->solidness = 0; // drawn separately, makes no faces
	}
	else // old style
	{
		f->solidness = 1;

		TileSpec t;
		if(g_texturesource)
			t.texture = g_texturesource->getTexture("water.png");
		
		t.alpha = WATER_ALPHA;
		t.material_type = MATERIAL_ALPHA_VERTEX;
		t.material_flags &= ~MATERIAL_FLAG_BACKFACE_CULLING;
		f->setAllTiles(t);
}
```

**Comment by TumeniNodes** [comment]
Did you try what I mentioned?
Comment out the special tiles...
Find water, dig 1 node down at the land edge about 6 nodes...
Allow 'water flowing' to begin and then block it with a solid node...
Then remove that solid node.

When I tried this, once I removed the solid node, for a split moment there is a face to water flowing with no texture associated with it, so it is replaced momentarily, with a dummy texture

**Comment by paramat** [comment]
Yes tried exactly that.
However the current code and the traced history suggest removing special tiles makes no difference.

**Comment by TumeniNodes** [comment]
It would be nice to see, as it makes no sense to have duplicate code which has no impact.

**Comment by paramat** [comment]
I'll make a PR soon.
